pypiserver/__main__.py

```diff
@@ -36,14 +42,132 @@ def init_logging(
         logger.addHandler(handler)
 
 
-def main(argv=None):
+class WsgiHandler(WSGIRequestHandler):
+    """A simple request handler to configure logging."""
+
+    # The default `FixedHandler` that bottle's `WSGIRefServer` uses does not
+    # log in a particularly predictable or configurable way. We'll pass this
+    # in to use instead.
+    def address_string(self) -> str:  # Prevent reverse DNS lookups please.
+        # This method copied directly from bottle's `FixedHandler` and
+        # maintained on the Chesterton's fence principle (i.e. I don't know
+        # why it's important, so I'm not going to get rid of it)
+        return self.client_address[0]
+
+    def log_message(
+        self, format: str, *args: t.Any  # pylint: disable=redefined-builtin
+    ) -> None:
+        """Log a message."""
+        # The log_message method on the `HttpRequestHandler` base class just
+        # writes directly to stderr. We'll use its same formatting, but pass
+        # it through the logger instead.
+        log.info(
+            "%s - - [%s] %s\n",
+            self.address_string(),
+            self.log_date_time_string(),
+            format % args,
+        )
+
+
+class ServerCheck:
+    """Methods to check for a server's presence in the import path."""
+
+    # Because we need to potentially override behavior for different servers
+    # by passing options into bottle's `run` method, and because those options
+    # aren't necessarily compatible between servers, we need to be able to guess
+    # which server bottle will pick if the user specifies the `auto` server
+    # type. These methods' names match the `__name__` property of the server
+    # adapters in bottle, so we can filter the adapters on the `AutoServer`
+    # adapter to figure out which ones are present. Since bottle uses the
+    # first server to successfully import from that same list, this should
+    # be sufficient to give us good certainty as to which server it will run.
+
+    # pylint: disable=missing-function-docstring
+    # pylint: disable=invalid-name
+    # pylint: disable=import-outside-toplevel
+    # pylint: disable=unused-import
+
+    @staticmethod
+    def WaitressServer() -> bool:
```

Comment:
wanna use a helper like?
```python
import importlib
def can_import(module):
    try:
        importlib.import_module(module)
        return True
    except ImportError:
        return False
```

Reply:
Yeah, this seems like a good idea. I thought about this, but leaned against it at the time because I wanted to be sure we were importing in the same way that bottle is, with direct statements. My understanding is that `importlib.import_module` will do the same thing though, so this should work too!